File: lib/src/main/java/io/xj/lib/meme/ParseUnique.java
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

package io.xj.lib.meme;

import java.util.List;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Meme Matcher for Unique-Memes
 * <p>
 * Parse any meme to test if it's valid, and extract its features
 * <p>
 * Artist can add `$MEME` so only one is chosen https://www.pivotaltracker.com/story/show/179078760
 */
class ParseUnique {
  static final Pattern rgx = Pattern.compile("^\\$(.+)$");
  final String body;
  final boolean isValid;

  ParseUnique(String raw) {
    Matcher matcher = rgx.matcher(raw);

    if (!matcher.find()) {
      body = null;
      isValid = false;
      return;
    }

    body = matcher.group(1);
    if (java.util.Objects.isNull(body) || body.length() == 0) {
      isValid = false;
      return;
    }

    isValid = true;
  }

  public static ParseUnique fromString(String raw) {
    return new ParseUnique(raw);
  }

  public boolean isViolatedBy(ParseUnique target) {
    return isValid && target.isValid && Objects.equals(body, target.body);
  }

  public boolean isAllowed(List<ParseUnique> memes) {
    for (var meme : memes)
      if (isViolatedBy(meme)) return false;
    return true;
  }
}

File: engine/src/entities/meme/ParseUnique.cpp
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

#include "xjmusic/entities/meme/ParseUnique.h"

using namespace XJ;

ParseUnique::ParseUnique(const std::string &raw) {
  std::smatch matcher;
  isValid = std::regex_search(raw, matcher, rgx);

  if (!isValid) {
    body = "";
    isValid = false;
    return;
  }

  body = matcher[1].str();
  if (body.empty()) {
    isValid = false;
    return;
  }

  isValid = true;
}

ParseUnique ParseUnique::fromString(const std::string &raw) {
  return ParseUnique(raw);
}

bool ParseUnique::isViolatedBy(const ParseUnique &target) {
  return isValid && target.isValid && body == target.body;
}

bool ParseUnique::isAllowed(const std::vector<ParseUnique> &memes) {
  return std::all_of(memes.begin(), memes.end(), [this](const ParseUnique &meme) {
    return !isViolatedBy(meme);
  });
}
